blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,784 @@
+
+```julia-repl
+julia> mi, node = tree[:backedges, 1]
+MethodInstance for reduce_empty(::Function, ::Type{T} where T) => MethodInstance for reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber}) at depth 0 with 143 children
+
+julia> node.mi      # this is the MethodInstance that called `mi`
+MethodInstance for reduce_empty(::Base.BottomRF{typeof(max)}, ::Type{VersionNumber})
+
+julia> typeintersect(tree.method.sig, node.mi.specTypes)
+Union{}
+```
+
+This is is a consequence of the fact that Julia's compiler has chosen not to specialize the argument types, and `reduce_empty(::Function, ::Type{T} where T)` is broader than what could be determined from the caller.
+Thus, while it looks like a case of greater specificity, in fact this is more analogous to the "partial specialization" described below.
+
+Moving backward another step, we get to `sizeof(::Type{X}) where X<:FixedPoint`.
+Simply put, this looks like a method that we don't need; perhaps it dates from some confusion, or an era where perhaps it was necessary.
+So we've discovered an easy place where a developer could do something to productively decrease the number of invalidations, in this case by just deleting the method.
+
+You'll also notice one example where the new method is *less specific*.
+It is not clear why such methods should be invalidating, and this may be a Julia bug.
+
+### Partial specialization
+
+If you try
+
+```julia-repl
+julia> trees = invalidation_trees(@snoopr using StaticArrays)
+```
+
+you'll see a much longer output.
+A large number of invalidations derive from the fact that StaticArrays defines a method for `!=`, which invalidates the fallback definition
+
+```
+!=(x, y) = !(x == y)
+```
+
+Since such definitions account for hundreds of nominal invalidations, it would be well worth considering whether it is possible to delete the custom `!=` methods.
+For example, if they are purely for internal use you could modify each caller to
+use the default method.
+
+The vast majority of the rest appear to derive from ambiguities.
+However, one more interesting case we've not seen before is
+
+```julia-repl
+julia> tree = trees[end-7]
+insert unsafe_convert(::Type{Ptr{T}}, m::Base.RefValue{FA}) where {N, T, D, FA<:FieldArray{N,T,D}} in StaticArrays at /home/tim/.julia/packages/StaticArrays/mlIi1/src/FieldArray.jl:124 invalidated:
+   mt_backedges: signature Tuple{typeof(Base.unsafe_convert),Type{Ptr{_A}} where _A,Base.RefValue{_A} where _A} triggered MethodInstance for unsafe_convert(::Type{Ptr{Nothing}}, ::Base.RefValue{_A} where _A) (159 children) more specific
+```
+
+In this case, the signature that triggered the invalidation, `Base.unsafe_convert(::Type{Ptr{_A}} where _A, ::Base.RefValue{_A} where _A)`,
+has been only partially specified: it depends on a type parameter `_A`.
```

Comment:
Isn't this two separate `_A` type parameters?  Or is it actually a confusing printing of the diagonal dispatch `Base.unsafe_convert(::Type{Ptr{_A}}, ::Base.RefValue{_A}) where _A`?